### DISCUSSION TITLE
[Question] Can OBS record a single video file with multiple video tracks using different sources?

### DISCUSSION THREAD

**Discussion by gerardbosch** [obsproject/obs-studio#9948] [General]
Hi, I couldn't find a clear answer to this, so I try to ask here :)

Some video containers like MKV (not sure if MP4) support multiple video tracks for the same file. Would it be possible to record multiple sources as different video tracks in the resulting file?

### The goal

- I would like to record my whole desktop and I have 2 screens.

So I'd like to record both screens in a way that after the video is recorded, I can either play it and switch from one track to the other, or ideally, that the video player opens in 2 windows, where each window plays one track and I can lay it out side-by-side or put every window on each screen.

Any chance to achieve it either natively or scripting something from the shell? I'd also want to be able to automate and start the recording from command line with a script so it can run in background.

Thank you!

**P.S.** This idea could be extended to further sources, like third monitor or webcam as independent track.

---

**EDIT:**

I found a plugin that allows to record several sources: https://obsproject.com/forum/resources/source-record.1285/
susing this you can enable a filter on the second source to record on a separate file. Then using ffmpeg they can be joined as a multi-track video: `ffmpeg -i vid1 -i video2 -c copy -map 0 -map 1 dual-vid.mkv`

This may help as starting point to wrap and run everything it from a bash script.

**Comment by sorayuki** [comment]
It's possible. LibOBS's API can handle this. But it may have not got a Graphics User Interface to operate.
I think it's possible to have it done via a lua or python script that OBS can loaded.

**Reply by gerardbosch** [reply]
Hi! So, the first question is: Can OBS UI do it? (record 2 video tracks to a single output video) I don't find a way. Do I need to find an extension script to do this? Does it exist?

The next question would be how to configure it tu run from a terminal (run OBS + start/stop recording).

**Reply by sorayuki** [reply]
It seems that no OBS UI doing that.
I can't give a example for the lua script immediately, it takes a long time to deal with signals for obs-output object.

**Reply by gerardbosch** [reply]
Thanks for answering :) I'm not familiar with OBS scripting, but as per your comments I'm understanding that no such feature exists out-of-the-box, but that recording a 2-track video should be possible by implementing an OBS plugin/extension (a Lua script?), is that correct?

This looked a quite general use case, isn't anything already implemented to allow recording multi-track video?

**Reply by sorayuki** [reply]
I think python script / lua script / DLL plugin can implement it.

In the latest OBS source code (30.0), there is a new API
```
/**
 * Sets the current video encoder associated with this output,
 * required for encoded outputs.
 *
 * The idx parameter specifies the video encoder index.
 * Only used with outputs that have multiple video outputs (FFmpeg typically),
 * otherwise the parameter is ignored.
 */
EXPORT void obs_output_set_video_encoder2(obs_output_t *output,
					  obs_encoder_t *encoder, size_t idx);
```
indicating a multi-track video feature.

**Reply by gerardbosch** [reply]
I made some findings with a filter plugin and added an EDIT to original post :)

**Comment by Fenrirthviti** [comment]
This sounds more likely to be a support request, and we are not currently accepting support requests on GitHub Discussions. Make sure you read any [pinned posts](https://github.com/obsproject/obs-studio/discussions/5206) on discussion boards before opening a new discussion. Please use our [forums](https://obsproject.com/forum/) or [Discord](https://obsproject.com/discord/) for further assistance.

Thank you!

**Reply by gerardbosch** [reply]
Sorry for posting here 🙏, I realized later about the pinned post. I'll try to move the question to the forums, but I'm not very sure under which category this would fit as I don't see a general support, only operating system specific support —this is a general, OS-agnostic question.

**Reply by Fenrirthviti** [reply]
This ultimately boils down to a normal support question, and mostly, a feature request. This is something that is currently supported on the OBS backend, but as sorayuki mentioned it is not exposed to the UI currently. I don't believe it is something that is possible to implement via scripting, it would need to be a proper plugin, or wait until we have sorted out the proper UI for this kind of feature.

https://ideas.obsproject.com/posts/41/multiple-video-outputs-selective-recording-iso-recording

It's one of our most requested features, and we're making steady steps towards it.

**Reply by gerardbosch** [reply]
Nice, thanks for your kind answer :)

**Reply by sorayuki** [reply]
I tried writting a lua script to save a multi video track file this morning, and stuck with no output module support OBS_OUTPUT_MULTI_TRACK_VIDEO exists.
Reimplementing a muxer in lua is a large project, but not impossible.

**Reply by gerardbosch** [reply]
Is there an "official" ticket to track this feature request? I find it an interesting feature and would like to track it for my own use in case it gets added in the future. I'm not specifically interested in the UI, as my goal is to run recording from a terminal, bash script or hotkey.

As @Fenrirthviti stated: _"This is something that is currently supported on the OBS backend"_ —so I guess that sooner or later would be exposed to any user interface (CLI or UI). Just would like to find it out in case it gets implemented. Thank you folks!!

**Comment by KristjanLaane** [comment]
yes please to multiple video tracks in same file / container, not just multitrack audio output
